File: qp/api/serializers/projects.py
from django.utils.translation import gettext_lazy as _
from rest_framework import serializers

from qp.api.serializers.users import qpUsersSimpleSerializer

from qp.projects.models import qpProject


class qpProjectsCreateSerializer(serializers.ModelSerializer):

    class Meta:
        model = qpProject
        fields = ["id", "slug", "name", "caption", "description", "primary_color", "creator", "owner"]
        read_only_fields = ["id", "slug"]


class qpProjectsDetailSerializer(serializers.ModelSerializer):
    owner = qpUsersSimpleSerializer(source="owner.profile")

    class Meta:
        model = qpProject
        fields = ["id", "slug", "name", "initial", "caption", "description", "primary_color", "owner"]
        read_only_fields = ["id", "slug", "name", "initial", "caption", "description", "primary_color", "owner"]


class qpProjectsDetailOwnerSerializer(serializers.ModelSerializer):

    class Meta:
        model = qpProject
        fields = ["id", "slug", "name", "initial", "caption", "description", "primary_color", "creator", "owner", "created_at",]
        read_only_fields = ["id", "slug"]

File: app/src/types/projects.ts
import type { UploadRawFile } from "element-plus";
import type { TypeUserSimple } from "./users";

export interface TypeProject {
  name: string,
  slug: string,
  caption?: string,
  description?: string,
  initial: string,
  primary_color: string,
  icon: string|null,
  creator: string|null,
  owner: TypeUserSimple|null
}

export interface ProjectsCreateForm {
  name: string,
  caption: string,
  description: string,
  primary_color: string,
  secondary_color: string,
  icon: UploadRawFile|string|null,
  icon_file: string|null
}